### ISSUE TITLE
How to detect objects smaller than the min bbox size of 32px?

### ISSUE THREAD

**Issue by wassname** [fizyr/keras-retinanet#202]
This is not an issue sorry, just a question. How would you go about detecting objects that are smaller than the smallest anchor size of 32px? For example cars in satellite images or seals on a rock. It's a difficult question for me because this repo and the paper are set up to ignore the high resolution feature layer C2 and have a minimum bbox size of 32 pixels.

I'm thinking:

- upscale the images, which will slow everything down. This works for me.
- modify it to use C2-C4 instead of use C3-C5 (I tried this [here](https://github.com/fizyr/keras-retinanet/compare/master...wassname:p2_16pix) but haven't got it to train well)
- change the scaling and stride. I tried this but ran into many errors and I'm not sure it's a good idea without feeding in C2 for higher resolution information.
- other?

**Comment by yhenon** [comment]
The smallest anchor is size 32px, so we can detect objects with IOU > 0.5 to a 32x32px box, which means sqrt(0.5*32*32) = 22.6 pixels, so about *23x23px* (in the case of a square box, fully aligned with the anchor box). Detecting smaller objects will probably be very challenging. 

Of the options you suggested, (1) is probably the best. (2) should work in theory, but it introduces a lot of new anchors, all of which are really hard to train (because small objects are challenging). This might be hard to train (as you observed), but could be ok if you decrease the LR and the parameter gamma.

**Comment by yhenon** [comment]
I'm going to go ahead and close this actually, since it is not an issue. Please use the keras team slack (see readme for instructions) for further questions & discussion.

**Comment by wassname** [comment]
Thanks for the feedback. I'll put questions there. And if anyone comes across this, I've found also that (1) is the best option.

**Comment by ferakon** [comment]
Hi, I know this issue is closed no, but wanted to share my solution, as it might help others.

Rather than scaling up the images, which increases memory usage and train time, I actually just forced my bboxes to be a minimum size (I found 29x29px seems to get roughly best results). This probably only works because my small objects are rare / quite sparsely distributed across my images. Also, my downstream processing doesn't care that the border round my identified object is slightly larger than that typically produced by this model. Might not work with densely packed objects because of non-max suppression, or multi objects contributing their features...

One thing to remember here is that scaling up the bboxes as part of a script like I did, rather than by manual annotation means that they sometimes go off the edge of the image. I just moved the bbox up or down accordingly so that the min > 0 and max < img dimension (len of x or y).

Cheers,

T.
